### PR TITLE
Simplify build pipeline

### PR DESCRIPTION
**Pull request by birkskyum**
We can remove the manual precompilation step (build-tsc) that generated /rollup/build/tsc. Blocked by #968

### CODE REVIEW

**Comment by HarelM** [comment]
Can we migrate totally to typescript plugin? Where are we still using the typescript output files?
If there's still a place I would recommend to migrate it first and then fully migrate to typescript plugin instead of adding it to the command line scripts...

**Comment by HarelM** [comment]
Also, since we have a build folder I would consider moving the rollup config there and move the bench folder under test folder?
Feels more organized this way (to me)...

**Comment by birkskyum** [comment]
> Can we migrate totally to typescript plugin? Where are we still using the typescript output files?

I'm unfortunately not sure I completely understand this question. If you ask if we still have references pointing to  /rollup/build/tsc/src rather than /src, then the answer is that all import statements are updated to use the ts files in /src directly now. That situation is not changed by this PR.

**Comment by HarelM** [comment]
Great! This means we can stop using the extra step of build-tsc and simply remove the watch flag from rollup config to always use the typescript plugin.
We should also add the typescript plugin to all rollup configs instead of adding them to the commandline argumets in the scripts in package.json.
I hope this is clearer now.
I would like to simplify the build as part of this PR.
I'm not saying this PR is not OK, just explaining how to improve it.

**Comment by birkskyum** [comment]
Just to understand the background - How come rollup only used the typescript() plugin when in watch mode?

**Comment by HarelM** [comment]
It was introduced after the migration was finished as I thought that there are issues with it when I did the migration.
It might be that it just didn't work as expected back then and I preferred to have something working.

**Comment by birkskyum** [comment]
I have tried to remove the watch() from the rollup config, but there are a few things that came up.

**Comment by HarelM** [comment]
Also it was easier to make the tests pass this way instead if forcing a full migration like we are in now. So it allowed us to do it one step at a time.

**Comment by HarelM** [comment]
Ahh, yes, the last part of the build simplification that I didn't manage to solve.
This part is replacing a file when using rollup.
This is basically mimicking the web worker bahavior.
I thought it should only be in the tests, but I couldn't make the tests pass without it.

**Comment by HarelM** [comment]
The way it should be as I see it, is that the new Worker code should be in src and there should be a hook to tell jest to use a worker mock like the implementation.
But it's not easy...
I'm surprised the browser tests and render tests are actually passing...?

**Comment by HarelM** [comment]
It took me a while to understand that there's a rollup plugin that uses a property defined in the package.json file to replace a file while building.
It's just the worst UX ever.
There were 3 files that were replaced and I managed to reduce it to only one file, which is better, but not perfect.

**Comment by birkskyum** [comment]
How do we test that this web worker replacement work?

**Comment by HarelM** [comment]
Not sure, maybe open the dev build artifacts in the browser and see that the code is not there (using npm run start-debug)?

File: bench/rollup_config_benchmarks.ts [hunk 1]
```diff
@@ -26,11 +26,10 @@ const replaceConfig = {
     'process.env.NODE_ENV': JSON.stringify('production')
 };
 
-const watch = process.env.ROLLUP_WATCH === 'true';
-const srcDir = watch ? '' : 'rollup/build/tsc/';
-const inputExt = watch ? 'ts' : 'js';
+const srcDir = '';
```
Comment (HarelM) [inline bench/rollup_config_benchmarks.ts:29]: There's no way to get different results like before, so I would suggest to remove these variables completely.

File: package.json [hunk 2]
```diff
@@ -145,22 +145,17 @@
...
-    "build-swc-bench": "swc bench -d rollup/build/tsc/bench -D -s",
-    "build-tsc": "npm run build-swc-src && npm run build-swc-bench",
-    "build-tsc-old": "tsc --outDir rollup/build/tsc",
-    "build-dev": "npm run build-tsc && rollup -c --environment BUILD:dev",
-    "watch-dev": "rollup -c --environment BUILD:dev --watch",
-    "build-prod": "npm run build-tsc && rollup -c --environment BUILD:production",
-    "build-prod-min": "npm run build-tsc && rollup -c --environment BUILD:production,MINIFY:true",
-    "build-csp": "rollup -c rollup.config.csp.js",
-    "build-query-suite": "rollup -c test/integration/rollup.config.test.js",
+    "build-dev": "rollup --configPlugin @rollup/plugin-typescript -c --environment BUILD:dev",
```
Comment (HarelM) [inline package.json:148]: No need for the plugin here as it is defined in the config, right?

File: rollup.config.ts [hunk 3]
```diff
@@ -1,17 +1,16 @@
...
+import banner from './build/banner';
 
-const {BUILD, MINIFY, ROLLUP_WATCH} = process.env;
+const {BUILD, MINIFY} = process.env;
 const minified = MINIFY === 'true';
-const watch = ROLLUP_WATCH === 'true';
-const srcDir = watch ? 'src' : 'rollup/build/tsc/src';
-const inputExt = watch ? 'ts' : 'js';
+const srcDir = 'src';
+const inputExt = 'ts';
```
Comment (HarelM) [inline rollup.config.ts:9]: Same here about the variables

File: rollup.config.ts [hunk 4]
```diff
@@ -20,7 +18,7 @@ export default [{
     // - rollup/build/maplibregl/shared.js: the set of modules that are dependencies of both the main module and the worker module
     //
     // This is also where we do all of our source transformations using the plugins.
-    input: [`${srcDir}/index.${inputExt}`, `${srcDir}/source/worker.${inputExt}`],
+    input: ['src/index.ts', 'src/source/worker.ts'],
```
Comment (HarelM) [inline rollup.config.ts:21]: The comment above this line probably needs to change as there's no more use for build-tsc.

**Comment by birkskyum** [comment]
> Not sure, maybe open the dev build artifacts in the browser and see that the code is not there (using npm run start-debug)?

I tried, and inn the start-debug, I'm still able to find the original web-worker script. I'm not that strong in debugging web workers still.

**Comment by birkskyum** [comment]
I managed to fix the style-spec build test, so now the only thing left is to make sure the web worker swap still works.

**Comment by birkskyum** [comment]
Another option is to use [sdom-worker](https://github.com/developit/jsdom-worker), and mock jest, leaving the source code clean.

**Comment by HarelM** [comment]
It would be great if we could avoid the replacement totally. I hope the referenced package can do that. If so it would be my preferred solution to remove the replacement and delete this extra code.

**Comment by birkskyum** [comment]
> It would be great if we could avoid the replacement totally. I hope the referenced package can do that. If so it would be my preferred solution to remove the replacement and delete this extra code.

So the web_worker.ts would instead of having the node stuff just look like this?

```ts
import maplibregl from '../index';

export default function () {
    return new Worker(maplibregl.workerUrl, {type:'module'});
}

```

**Comment by birkskyum** [comment]
@HarelM , now that #962  #968  are in, this is updated, and works - even benchmarks.

**Comment by HarelM** [comment]
COOL!!! Nice work mate! Including the web workers, right?

**Review by HarelM — APPROVED** [review]

**Comment by birkskyum** [comment]
> Nice work mate

Including the web workers!

**Comment by birkskyum** [comment]
The render tests are not on jest yet - still on harness as the last piece, but they will get there. I tried to move stub_loader to ts, but it turned out to be breaking change, so that file is still .js, but we are pretty much typescript throughout.

**Pull request merged**